src/Compiler/CompilerOptions.hpp

```diff
@@ -48,57 +50,67 @@ typedef enum {
 
 extern const std::string modelSizeStr[];
 
-// Options for onnx-mlir only.
-extern llvm::cl::OptionCategory OnnxMlirOptions;
 // Common options shared between onnx-mlir and onnx-mlir-opt.
 extern llvm::cl::OptionCategory OnnxMlirCommonOptions;
-
-extern llvm::cl::opt<bool> invokeOnnxVersionConverter;
-extern llvm::cl::opt<bool> preserveLocations;
-extern llvm::cl::opt<bool> printIR;
-extern llvm::cl::opt<bool> preserveBitcode;
-extern llvm::cl::opt<bool> preserveLLVMIR;
-extern llvm::cl::opt<bool> preserveMLIR;
-extern llvm::cl::opt<bool> useOnnxModelTypes;
-extern llvm::cl::opt<int> repeatOnnxTransform;
-extern llvm::cl::opt<std::string> shapeInformation;
-extern llvm::cl::opt<onnx_mlir::OptLevel> OptimizationLevel;
-extern llvm::cl::opt<std::string> customEnvFlags;
-extern llvm::cl::opt<std::string> mtriple;
-extern llvm::cl::opt<std::string> mcpu;
-extern llvm::cl::opt<std::string> march;
-extern llvm::cl::opt<ModelSize> modelSize;
-extern llvm::cl::opt<bool> storeConstantsToFile;
-extern llvm::cl::opt<float> constantsToFileSingleThreshold;
-extern llvm::cl::opt<float> constantsToFileTotalThreshold;
-extern llvm::cl::list<onnx_mlir::accel::Accelerator::Kind> maccel;
-extern llvm::cl::opt<bool> VerboseOutput;
-extern llvm::cl::list<std::string> Xopt;
-extern llvm::cl::list<std::string> Xllc;
-extern llvm::cl::opt<std::string> mllvm;
-extern llvm::cl::opt<bool> verifyInputTensors;
-extern llvm::cl::opt<bool> allowSorting;
-extern llvm::cl::opt<std::string> reportHeapBefore;
-extern llvm::cl::opt<std::string> reportHeapAfter;
-extern llvm::cl::opt<InstrumentStages> instrumentStage;
-extern llvm::cl::opt<std::string> instrumentOps;
-extern llvm::cl::bits<InstrumentActions> instrumentControlBits;
-extern llvm::cl::opt<bool> instrumentONNXSignature;
-extern llvm::cl::opt<std::string> ONNXOpStats;
-extern llvm::cl::opt<bool> enableMemoryBundling;
-extern llvm::cl::opt<int> onnxOpTransformThreshold;
-extern llvm::cl::opt<bool> onnxOpTransformReport;
-extern llvm::cl::opt<int> onnxConstPropExpansionBound;
-extern llvm::cl::opt<bool> enableParallel;
-extern llvm::cl::opt<bool> disableSimdOption;
-extern llvm::cl::opt<bool> enableSimdDataLayout;
-extern llvm::cl::opt<bool> enableONNXHybridPass;
-extern llvm::cl::list<std::string> functionsToDecompose;
-extern llvm::cl::opt<std::string> modelTag;
-extern llvm::cl::opt<bool> enableConvOptPass;
-extern llvm::cl::list<std::string> extraLibPaths;
-extern llvm::cl::list<std::string> extraLibs;
-extern llvm::cl::opt<ProfileIRs> profileIR;
+// Options for onnx-mlir only.
+extern llvm::cl::OptionCategory OnnxMlirOptions;
+// Options for onnx-mlir-opt only.
+extern llvm::cl::OptionCategory OnnxMlirOptOptions;
+
+// Options known to onnx-mlir and/or onnx-mlir-opt
```

Comment:
I just tried and it appears that the option belonging to the other command is silently ignored. #ifdef at compile time doesn't work because CompilerOptions.cpp is shared by both commands so at compile time you always have to include both sets of options. Maybe there is some way to partition the options using namespace but I'm not entirely familiar with that. Perhaps we can address the issue in a followup PR.

Reply:
I see that LLVM/MLIR organizes options by defining `registerXXXOptions` for different categories of options, and register ones when needed.

For example, in `mlir-opt` command:  https://github.com/llvm/llvm-project/blob/main/mlir/lib/Tools/mlir-opt/MlirOptMain.cpp#L491-L497 
```
 // Register any command line options.
  MlirOptMainConfig::registerCLOptions(registry);
  registerAsmPrinterCLOptions();
  registerMLIRContextCLOptions();
  registerPassManagerCLOptions();
  registerDefaultTimingManagerCLOptions();
  tracing::DebugCounter::registerCLOptions();
  ```
  
while in `mlir-reduce` command: https://github.com/llvm/llvm-project/blob/main/mlir/lib/Tools/mlir-translate/MlirTranslateMain.cpp#L89C2-L93C59
```
  registerAsmPrinterCLOptions();
  registerMLIRContextCLOptions();
  registerTranslationCLOptions();
  registerDefaultTimingManagerCLOptions();
```

Haven't dug into details to see how it does.

Reply:
It turns out that there is a fairly simple way to make onnx-mlir not be able to use onnx-mlir-opt's options and vise versa. We call `llvm::cl::HideUnrelatedOptions` to make `--help` not print unrelated options. Those options are marked with `llvm::cl::ReallyHidden` flag. We can use `llvm::cl::getRegisteredOptions()` to get a map of the options and remove those that are marked with the `llvm::cl::ReallyHidden` flag. I've tried it and it actually works. I will update the PR with the fix.

To make the options fully dynamically constructed like what LLVM/MLIR does above likely requires major rework of the options code. Perhaps we can leave that as something for the future.